File: frontend/src/api/mcp.ts
/**
 * MCP server API — 对接 backend/app/api/routes/mcp/*
 *
 * 路径前缀 `/mcp-servers`，加上 axios baseURL `/api/v1`。
 */

import { del, get, post, put } from '@/request'
import type { MCPServer, MCPTransport } from '@/types'

export interface MCPServerCreatePayload {
  name: string
  server_url: string
  transport?: MCPTransport
  /** 自定义请求头（含鉴权 token，明文提交，后端加密存储） */
  headers?: Record<string, string>
  description?: string
  enabled?: boolean
}

export interface MCPServerUpdatePayload {
  name?: string
  server_url?: string
  transport?: MCPTransport
  /** 新请求头，留空（不传）则不改 */
  headers?: Record<string, string>
  description?: string
  enabled?: boolean
}

/** 列出当前用户的 MCP server。 */
export function listMCPServers() {
  return get<MCPServer[]>('/mcp-servers')
}

/** 创建 MCP server。走 silent 由表单层 toast，方便显示后端具体原因。 */
export function createMCPServer(payload: MCPServerCreatePayload) {
  return post<MCPServer>('/mcp-servers', payload, { silent: true })
}

/** 更新 MCP server（部分字段；启用开关切换也走这里）。 */
export function updateMCPServer(id: string, payload: MCPServerUpdatePayload) {
  return put<MCPServer>(`/mcp-servers/${id}`, payload)
}

/** 删除 MCP server。 */
export function deleteMCPServer(id: string) {
  return del<null>(`/mcp-servers/${id}`)
}

export interface MCPTestConnectionPayload {
  server_url: string
  transport?: MCPTransport
  headers?: Record<string, string>
}

export interface MCPToolBrief {
  name: string
  description: string
}

export interface MCPTestConnectionResult {
  success: boolean
  tool_count: number
  tools: MCPToolBrief[]
  error: string | null
}

/**
 * 测试连接：用配置直接试连 MCP server，返回连通性 + 发现的工具列表。
 *
 * 后端把「连不上」当正常业务结果（success=false + error），不抛 HTTP 错误；
 * 走 silent 由弹窗内联展示结果。
 */
export function testMCPConnection(payload: MCPTestConnectionPayload) {
  return post<MCPTestConnectionResult>('/mcp-servers/test', payload, {
    silent: true,
  })
}

File: backend/app/api/routes/mcp/mcp.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends

from app.core.depends import get_current_user
from app.core.http import ResponseModel, success
from app.models.user import User
from app.schemas.mcp import (
    MCPServerCreate,
    MCPServerOut,
    MCPServerUpdate,
    MCPTestConnectionIn,
    MCPTestConnectionOut,
    MCPToolBrief,
)
from app.services.mcp import MCPServerService, get_mcp_server_service

router = APIRouter(prefix="/mcp-servers", tags=["mcp-servers"])

CurrentUserDep = Annotated[User, Depends(get_current_user)]
MCPServerServiceDep = Annotated[MCPServerService, Depends(get_mcp_server_service)]


@router.post("", summary="创建 MCP server")
async def create_mcp_server(
    data: MCPServerCreate,
    current_user: CurrentUserDep,
    svc: MCPServerServiceDep,
) -> ResponseModel[MCPServerOut]:
    server = await svc.create(current_user, data)
    return success(data=MCPServerOut.model_validate(server), message="创建成功")


@router.post("/test", summary="测试 MCP server 连接")
async def test_mcp_connection(
    data: MCPTestConnectionIn,
    current_user: CurrentUserDep,
    svc: MCPServerServiceDep,
) -> ResponseModel[MCPTestConnectionOut]:
    ok, tools, error = await svc.test_connection(
        data.server_url, data.transport, data.headers
    )
    return success(
        data=MCPTestConnectionOut(
            success=ok,
            tool_count=len(tools),
            tools=[
                MCPToolBrief(name=t.name, description=t.description or "")
                for t in tools
            ],
            error=error,
        )
    )


@router.get("", summary="列出自己的 MCP server")
async def list_mcp_servers(
    current_user: CurrentUserDep,
    svc: MCPServerServiceDep,
) -> ResponseModel[list[MCPServerOut]]:
    servers = await svc.list_own(current_user)
    return success(data=[MCPServerOut.model_validate(s) for s in servers])


@router.get("/{server_id}", summary="MCP server 详情")
async def get_mcp_server(
    server_id: UUID,
    current_user: CurrentUserDep,
    svc: MCPServerServiceDep,
) -> ResponseModel[MCPServerOut]:
    server = await svc.get_by_id(current_user, server_id)
    return success(data=MCPServerOut.model_validate(server))


@router.put("/{server_id}", summary="更新 MCP server")
async def update_mcp_server(
    server_id: UUID,
    data: MCPServerUpdate,
    current_user: CurrentUserDep,
    svc: MCPServerServiceDep,
) -> ResponseModel[MCPServerOut]:
    server = await svc.update(current_user, server_id, data)
    return success(data=MCPServerOut.model_validate(server), message="更新成功")


@router.delete("/{server_id}", summary="删除 MCP server")
async def delete_mcp_server(
    server_id: UUID,
    current_user: CurrentUserDep,
    svc: MCPServerServiceDep,
) -> ResponseModel[None]:
    await svc.delete(current_user, server_id)
    return success(message="删除成功")
